File: python/src/toolnexus/toolkit.py
"""Toolkit — aggregates dynamic MCP tools + the skill tool + custom tools into a
single uniform list, with provider adapters and a single ``execute()`` router.

The MCP SDK is async, so the toolkit is async: build it with the
:func:`create_toolkit` factory and manage its lifetime with ``async with`` (or
call :meth:`Toolkit.close`). Mirrors the JS reference (``js/src/toolkit.ts``).
"""
from __future__ import annotations

import asyncio
import inspect
import sys
from typing import TYPE_CHECKING, Any, Optional

from .a2a import Agent, agent_tools, parse_agents_config
from .adapters import to_anthropic, to_gemini, to_openai
from .builtin import BuiltinsConfig, select_builtins
from .mcp_serve import (
    MCPServeConfig,
    OnCall,
)
from .mcp_source import McpSource, load_mcp_with_context
from .serve import A2AConfig, OnTask, ServeHandle, start_a2a_server
from .skill import SkillDef, SkillProvider, SkillSource, load_skills
from .types import McpStatus, Tool, ToolContext, ToolResult

if TYPE_CHECKING:  # typing-only, avoid a runtime import cycle with client.py
    from .client import Client


class Toolkit:
    def __init__(
        self,
        mcp: Optional[McpSource],
        skill: Optional[SkillSource],
        builtins: list[Tool],
        agents: list[Tool],
        extra_tools: list[Tool],
        a2a_config: Optional[A2AConfig] = None,
        mcp_server_config: Optional[MCPServeConfig] = None,
        disable_tools: Optional[list[str]] = None,
    ) -> None:
        self._mcp = mcp
        self._skill = skill
        # A2A profile read from a top-level `a2a` config block (serve() falls back to this).
        self._a2a_config = a2a_config
        # MCP serve profile read from a top-level `mcpServer` (singular) config block
        # — distinct from the client-side `mcpServers`. serve() prefers its inline
        # `mcp` over this.
        self._mcp_server_config = mcp_server_config
        self._by_name: dict[str, Tool] = {}
        # Builtins are the lowest-precedence source: a host extra_tools entry with
        # the same name shadows a builtin (SPEC §4). Drop shadowed builtins up
        # front, then apply the normal first-wins dedupe for MCP/skill/agents/extras.
        extra_names = {t.name for t in extra_tools}
        active_builtins = [b for b in builtins if b.name not in extra_names]
        all_tools: list[Tool] = []
        if mcp is not None:
            all_tools.extend(mcp.tools)
        if skill is not None:
            all_tools.append(skill.tool)
        all_tools.extend(active_builtins)
        all_tools.extend(agents)
        all_tools.extend(extra_tools)
        # DisableTools drops tools by their FINAL exposed name across every source
        # (MCP `server_tool`, builtins, native, a2a), applied at aggregation. Composes
        # with the per-server/per-skill map filters. Mirrors Go's DisableTools.
        disabled = set(disable_tools or ())
        for t in all_tools:
            if t.name in disabled:
                continue
            if t.name in self._by_name:
                print(
                    f'[toolnexus] duplicate tool name "{t.name}" — keeping first',
                    file=sys.stderr,
                )
                continue
            self._by_name[t.name] = t

    @classmethod
    async def create(
        cls,
        mcp_config: Optional[str | dict[str, Any]] = None,
        skills_dir: Optional[str | list[str]] = None,
        skills: Optional[list[SkillDef]] = None,
        skill_provider: Optional[SkillProvider] = None,
        skills_filter: Optional[dict[str, bool]] = None,
        skill_sample_limit: int = 0,
        extra_tools: Optional[list[Tool]] = None,
        builtins: Optional[BuiltinsConfig] = None,
        agents: Optional[list[Agent]] = None,
        wait_for: Optional[Any] = None,
        disable_tools: Optional[list[str]] = None,
        disable_skills: Optional[list[str]] = None,
        cancel: Optional[Any] = None,
    ) -> "Toolkit":
        # ``wait_for`` (§10): when set, connected MCP servers may elicit input from the
        # human mid-``tools/call``, bridged onto this resolver (form→kind:"input",
        # URL→kind:"authorization"). Typically the same function passed to the client.
        # Omit ⇒ MCP elicitation is not advertised.
        # ``cancel`` (§2 Gap 3): an optional asyncio.Event; when set it aborts the whole
        # MCP load (the Python analogue of a cancelled parent context).
        mcp = (
            await load_mcp_with_context(mcp_config, wait_for=wait_for, cancel=cancel)
            if mcp_config is not None
            else None
        )
        # Skills come from directories, in-memory data, and/or a lazy provider
        # (§3, S1). The provider is resolved here (create() is async) and merged
        # with the data list; a provider failure is isolated so other sources
        # still load.
        skill = None
        if skills_dir is not None or skills is not None or skill_provider is not None:
            provider_defs: list[SkillDef] = []
            if skill_provider is not None:
                try:
                    result = skill_provider()
                    if inspect.isawaitable(result):
                        result = await result
                    provider_defs = list(result)
                except Exception as e:  # noqa: BLE001 — provider failure is isolated
                    print(f"[toolnexus] skill provider failed: {e}", file=sys.stderr)
                    provider_defs = []
            data_defs = list(skills or []) + provider_defs
            # DisableSkills is sugar over a drop-list: fold each name into the filter as
            # False (without overriding an explicit skills_filter entry). Mirrors Go.
            eff_filter = skills_filter
            if disable_skills:
                merged = dict(skills_filter or {})
                for n in disable_skills:
                    merged.setdefault(n, False)
                eff_filter = merged
            skill = load_skills(
                skills_dir,
                skills=data_defs or None,
                filter=eff_filter,
                sample_limit=skill_sample_limit,
            )
        # The toggle comes from the `builtins` option, or a top-level `builtins`
        # key on a parsed config dict — same precedence as MCP's is_enabled.
        builtins_cfg = builtins
        if (
            builtins_cfg is None
            and isinstance(mcp_config, dict)
            and "builtins" in mcp_config
        ):
            builtins_cfg = mcp_config["builtins"]
        builtin_tools = select_builtins(builtins_cfg)

        # Remote A2A agents come from the `agents=[...]` option plus a top-level
        # `agents` block on a parsed config dict (mirrors mcpServers). Each is
        # resolved to its skill tools; a failing agent never breaks the toolkit.
        agent_descriptors: list[Agent] = list(agents or [])
        if isinstance(mcp_config, dict) and "agents" in mcp_config:
            agent_descriptors.extend(parse_agents_config(mcp_config["agents"]))

        async def _resolve(ag: Agent) -> list[Tool]:
            try:
                return await agent_tools(ag)
            except Exception as e:  # noqa: BLE001 — one bad agent never breaks the toolkit
                print(
                    f'[toolnexus] A2A agent "{ag.card}" failed: {e}',
                    file=sys.stderr,
                )
                return []

        agent_tool_lists = await asyncio.gather(*(_resolve(a) for a in agent_descriptors))
        agent_tools_flat = [t for lst in agent_tool_lists for t in lst]

        # A2A inbound profile: a top-level `a2a` block on a parsed config dict
        # (mirrors builtins/agents). serve() prefers its inline `a2a` over this.
        a2a_config: Optional[A2AConfig] = None
        if isinstance(mcp_config, dict) and "a2a" in mcp_config:
            a2a_config = mcp_config["a2a"]

        # MCP inbound profile: a top-level `mcpServer` (singular) block — distinct
        # from the client-side `mcpServers` block. serve() prefers its inline `mcp`
        # over this.
        mcp_server_config: Optional[MCPServeConfig] = None
        if isinstance(mcp_config, dict) and "mcpServer" in mcp_config:
            mcp_server_config = mcp_config["mcpServer"]

        return cls(
            mcp,
            skill,
            builtin_tools,
            agent_tools_flat,
            extra_tools or [],
            a2a_config,
            mcp_server_config,
            disable_tools=disable_tools,
        )

    def register(self, *tools: Tool) -> "Toolkit":
        """Add native/http/custom tools at runtime. First-name-wins; warn on dup.

        Returns ``self`` so calls can be chained.
        """
        for t in tools:
            if t.name in self._by_name:
                print(
                    f'[toolnexus] duplicate tool name "{t.name}" — keeping first',
                    file=sys.stderr,
                )
                continue
            self._by_name[t.name] = t
        return self

    async def add_agent(
        self,
        agent_or_card_url: "Agent | str",
        *,
        headers: Optional[dict[str, str]] = None,
        timeout: Optional[int] = None,
        poll_every: Optional[int] = None,
    ) -> "Toolkit":
        """Fetch a remote A2A agent's card at runtime and register its skills as
        tools. Accepts an :class:`Agent` descriptor or a bare card URL.
        First-name-wins dedupe. Returns ``self`` so calls can be chained.
        """
        if isinstance(agent_or_card_url, str):
            ag = Agent(
                card=agent_or_card_url,
                headers=headers,
                timeout=timeout,
                poll_every=poll_every,
            )
        else:
            ag = agent_or_card_url
        tools = await agent_tools(ag)
        return self.register(*tools)

    async def serve(
        self,
        addr: str,
        *,
        client: Optional["Client"] = None,
        a2a: Optional[A2AConfig] = None,
        on_task: Optional[OnTask] = None,
        mcp: Optional[MCPServeConfig] = None,
        on_call: Optional[OnCall] = None,
    ) -> ServeHandle:
        """Serve this toolkit as an agent over HTTP. When the ``a2a`` profile is
        present (inline, or a top-level ``a2a`` config block the toolkit was built
        from), it mounts the A2A Agent Card (``/.well-known/agent-card.json``, built
        from skills) and a JSON-RPC endpoint (``SendMessage`` + ``GetTask``) fulfilled
        by the client. A message's A2A ``contextId`` keys the conversation via
        ``client.ask``, so a peer's turns are remembered through the client's
        ConversationStore. When the ``mcp`` profile is present (inline, or a top-level
        ``mcpServer`` config block), a streamable-HTTP MCP server exposing the
        toolkit's unified tools is co-mounted at ``POST /mcp``. When both are absent,
        no routes are mounted. Returns a stoppable handle.
        """
        cfg = a2a if a2a is not None else self._a2a_config
        mcp_cfg = mcp if mcp is not None else self._mcp_server_config
        skills = list(self._skill.skills.values()) if self._skill is not None else []

        async def run_task(text: str, context_id: Optional[str] = None) -> Any:
            # contextId keys the conversation via client.ask so a peer's turns are
            # remembered through the client's ConversationStore; absent ⇒ stateless run.
            if context_id:
                return await client.ask(text, self, id=context_id)  # type: ignore[union-attr]
            return await client.run(text, self)  # type: ignore[union-attr]

        return await start_a2a_server(
            addr=addr,
            a2a=cfg,
            skills=skills,
            run_task=run_task,
            on_task=on_task,
            loop=asyncio.get_running_loop(),
            mcp=mcp_cfg,
            mcp_tools=self.tools() if mcp_cfg else None,
            on_call=on_call,
        )

    def tools(self) -> list[Tool]:
        return list(self._by_name.values())

    def get(self, name: str) -> Optional[Tool]:
        return self._by_name.get(name)

    async def execute(
        self,
        name: str,
        args: Optional[dict[str, Any]] = None,
        ctx: Optional[ToolContext] = None,
    ) -> ToolResult:
        tool = self._by_name.get(name)
        if tool is None:
            return ToolResult(output=f"Unknown tool: {name}", is_error=True)
        return await tool.execute(args or {}, ctx)

    def skills_prompt(self) -> str:
        """Markdown skill catalog for the system prompt."""
        return self._skill.prompt() if self._skill is not None else ""

    def mcp_status(self) -> dict[str, McpStatus]:
        return dict(self._mcp.status) if self._mcp is not None else {}

    def to_openai(self) -> list[dict[str, Any]]:
        return to_openai(self.tools())

    def to_anthropic(self) -> list[dict[str, Any]]:
        return to_anthropic(self.tools())

    def to_gemini(self) -> list[dict[str, Any]]:
        return to_gemini(self.tools())

    async def close(self) -> None:
        if self._mcp is not None:
            await self._mcp.close()

    async def __aenter__(self) -> "Toolkit":
        return self

    async def __aexit__(self, *exc: Any) -> None:
        await self.close()


async def create_toolkit(
    mcp_config: Optional[str | dict[str, Any]] = None,
    skills_dir: Optional[str | list[str]] = None,
    skills: Optional[list[SkillDef]] = None,
    skill_provider: Optional[SkillProvider] = None,
    skills_filter: Optional[dict[str, bool]] = None,
    skill_sample_limit: int = 0,
    extra_tools: Optional[list[Tool]] = None,
    builtins: Optional[BuiltinsConfig] = None,
    agents: Optional[list[Agent]] = None,
    wait_for: Optional[Any] = None,
    disable_tools: Optional[list[str]] = None,
    disable_skills: Optional[list[str]] = None,
    cancel: Optional[Any] = None,
) -> Toolkit:
    """Async factory. The returned Toolkit is also an async context manager.

    Built-in tools (§4A) are on by default; pass ``builtins=False`` /
    ``{"disabled": True}`` / ``{"enabled": False}`` to turn the whole source off.
    Remote A2A agents (§7A) are supplied via ``agents=[agent(...), ...]`` and/or a
    top-level ``agents`` block on a parsed config dict.

    ``disable_tools`` drops tools by their final exposed name across every source;
    ``disable_skills`` drops skills by name from the ``skill`` catalog (both compose
    with the per-server/per-skill map filters). ``cancel`` is an optional
    :class:`asyncio.Event` propagated into the MCP load (§2 Gap 3).
    """
    return await Toolkit.create(
        mcp_config=mcp_config,
        skills_dir=skills_dir,
        skills=skills,
        skill_provider=skill_provider,
        skills_filter=skills_filter,
        skill_sample_limit=skill_sample_limit,
        extra_tools=extra_tools,
        builtins=builtins,
        agents=agents,
        wait_for=wait_for,
        disable_tools=disable_tools,
        disable_skills=disable_skills,
        cancel=cancel,
    )

File: elixir/lib/toolnexus/toolkit.ex
defmodule Toolnexus.Toolkit do
  @moduledoc """
  Toolkit — aggregates dynamic MCP tools + the skill tool + built-ins + remote
  A2A agents + custom tools into a single uniform list, with provider adapters
  and a single `execute/4` router (SPEC §4).

  Assembly is deterministic: (1) drop any builtin whose name a host
  `extra_tools` entry provides (a host can shadow a builtin); (2) concatenate
  the remaining sources in the order **MCP → skill → builtin → agents →
  extra_tools**; (3) dedupe by name, **first wins**. `disable_tools` drops
  tools by their final exposed name across every source.

  The struct exposes `:tools` and `:prompt` directly, so it plugs straight
  into `Toolnexus.Client.run/4`.
  """

  require Logger

  alias Toolnexus.{A2a, Adapters, Builtin, Context, Mcp, Skill, Tool, ToolResult}

  defstruct tools: [],
            prompt: "",
            by_name: %{},
            mcp: nil,
            skill: nil,
            a2a_config: nil,
            mcp_server_config: nil

  @type t :: %__MODULE__{
          tools: [Tool.t()],
          prompt: String.t(),
          by_name: %{String.t() => Tool.t()},
          mcp: Mcp.Source.t() | nil,
          skill: Skill.Source.t() | nil,
          a2a_config: map() | nil,
          mcp_server_config: map() | nil
        }

  @doc """
  Build a Toolkit. Options (keyword list or map):

    * `:mcp_config` — path to an MCP config file, raw JSON, or a parsed map
    * `:skills_dir` — one or more skill roots
    * `:skills` — skills supplied as data, bypassing the filesystem (§3, S1)
    * `:skill_provider` — 0-arity fun lazily supplying data skills (§3, S1)
    * `:skills_filter` — per-agent skill allowlist keyed on name (§3, S2)
    * `:skill_sample_limit` — sibling-file cap: 0 ⇒ default 10, n>0 ⇒ cap,
      -1 ⇒ omit `<skill_files>` (§3, S5)
    * `:disable_tools` — drop tools by their FINAL exposed name across every
      source; composes with the per-server/per-skill filters
    * `:disable_skills` — drop skills by name from the `skill` catalog (sugar
      over a drop-list in `:skills_filter`)
    * `:extra_tools` — your own custom tools, always added to the toolkit
    * `:builtins` — built-in tools toggle (§4A). Default ON. When absent, a
      top-level `builtins` key on a parsed config map is consulted
    * `:agents` — remote A2A agents (§7A); a top-level `agents` config block
      is merged in as well
    * `:wait_for` — §10 host resolver; bridges MCP elicitation
    * `:deadline` (ms) — bounds the MCP load (§2 Gap 3)
  """
  @spec build(keyword() | map()) :: t()
  def build(opts) do
    opts = Map.new(opts)

    # Reserved sibling keys (builtins / agents / a2a / mcpServer) are read off
    # a PARSED config map only — a path/raw-JSON config is the MCP source's
    # business alone (JS parity: `typeof mcpConfig === "object"`).
    raw =
      case opts[:mcp_config] do
        cfg when is_map(cfg) -> cfg |> Jason.encode!() |> Jason.decode!()
        _ -> %{}
      end

    mcp =
      if opts[:mcp_config] != nil,
        do: Mcp.load(opts[:mcp_config], wait_for: opts[:wait_for], deadline: opts[:deadline])

    skill = load_skill_source(opts)

    # The toggle comes from the `builtins` option, or a top-level `builtins`
    # key on a parsed config map — same precedence as MCP's isEnabled.
    builtins_cfg = if Map.has_key?(opts, :builtins), do: opts[:builtins], else: raw["builtins"]
    builtins = Builtin.load(builtins_cfg)

    # Remote A2A agents come from the `agents` option plus a top-level
    # `agents` config block (mirrors mcpServers). Each is resolved to its
    # skill tools; a failing agent is isolated and never breaks the toolkit.
    agent_descriptors =
      Enum.map(opts[:agents] || [], &Map.new/1) ++ A2a.parse_agents_config(raw["agents"])

    agent_tools =
      Enum.flat_map(agent_descriptors, fn ag ->
        try do
          A2a.agent_tools(ag)
        rescue
          e ->
            Logger.warning(~s([toolnexus] A2A agent "#{ag[:card]}" failed: #{Exception.message(e)}))
            []
        end
      end)

    extra_tools = opts[:extra_tools] || []

    # Builtins are the lowest-precedence source: a host extra_tools entry with
    # the same name shadows a builtin (SPEC §4). Drop shadowed builtins up
    # front, then apply the normal first-wins dedupe.
    extra_names = MapSet.new(extra_tools, & &1.name)
    active_builtins = Enum.reject(builtins, &MapSet.member?(extra_names, &1.name))

    all =
      ((mcp && mcp.tools) || []) ++
        if(skill, do: [skill.tool], else: []) ++ active_builtins ++ agent_tools ++ extra_tools

    # disable_tools drops tools by their final exposed name across every source.
    disabled = MapSet.new(opts[:disable_tools] || [])

    {order_rev, by_name} =
      all
      |> Enum.reject(&MapSet.member?(disabled, &1.name))
      |> Enum.reduce({[], %{}}, &dedupe/2)

    %__MODULE__{
      tools: Enum.reverse(order_rev),
      prompt: if(skill, do: skill.prompt, else: ""),
      by_name: by_name,
      mcp: mcp,
      skill: skill,
      a2a_config: raw["a2a"],
      mcp_server_config: raw["mcpServer"]
    }
  end

  defp dedupe(tool, {order, by_name}) do
    if Map.has_key?(by_name, tool.name) do
      Logger.warning(~s([toolnexus] duplicate tool name "#{tool.name}" — keeping first))
      {order, by_name}
    else
      {[tool | order], Map.put(by_name, tool.name, tool)}
    end
  end

  # Skills come from directories, in-memory data, and/or a lazy provider (§3,
  # S1). `disable_skills` is sugar over a drop-list: fold each name into the
  # filter as false (without overriding an explicit skills_filter entry).
  defp load_skill_source(opts) do
    if opts[:skills_dir] != nil or opts[:skills] != nil or opts[:skill_provider] != nil do
      filter = opts[:skills_filter]

      filter =
        case opts[:disable_skills] || [] do
          [] ->
            filter

          names ->
            Enum.reduce(names, filter || %{}, fn n, acc ->
              if Map.has_key?(acc, n), do: acc, else: Map.put(acc, n, false)
            end)
        end

      Skill.load(
        dirs: opts[:skills_dir],
        skills: opts[:skills],
        provider: opts[:skill_provider],
        filter: filter,
        sample_limit: opts[:skill_sample_limit]
      )
    end
  end

  @doc "All tools (mcp tools + skill tool + builtins + agents + extras), in insertion order."
  @spec tools(t()) :: [Tool.t()]
  def tools(%__MODULE__{tools: tools}), do: tools

  @doc "A tool by name, or `nil`."
  @spec get(t(), String.t()) :: Tool.t() | nil
  def get(%__MODULE__{by_name: by_name}, name), do: Map.get(by_name, name)

  @doc """
  Add tools at runtime (native, http, or any custom Tool). First name wins;
  duplicates are skipped with a warning. Returns the updated toolkit.
  """
  @spec register(t(), [Tool.t()] | Tool.t()) :: t()
  def register(%__MODULE__{} = tk, tools) do
    {order_rev, by_name} =
      Enum.reduce(List.wrap(tools), {Enum.reverse(tk.tools), tk.by_name}, &dedupe/2)

    %{tk | tools: Enum.reverse(order_rev), by_name: by_name}
  end

  @doc """
  Fetch a remote A2A agent's card at runtime and register its skills as tools.
  Accepts an Agent descriptor (map/keyword) or a bare card URL (with optional
  `opts` for headers/timeout/poll_every). First-name-wins dedupe.
  """
  @spec add_agent(t(), map() | keyword() | String.t(), keyword() | map()) :: t()
  def add_agent(tk, agent_or_card_url, opts \\ [])

  def add_agent(%__MODULE__{} = tk, card_url, opts) when is_binary(card_url),
    do: add_agent(tk, Map.put(Map.new(opts), :card, card_url), [])

  def add_agent(%__MODULE__{} = tk, ag, _opts),
    do: register(tk, A2a.agent_tools(A2a.agent(ag)))

  @doc "Route to the right tool and run it. Unknown name ⇒ an error ToolResult."
  @spec execute(t(), String.t(), map(), Context.t() | nil) :: ToolResult.t()
  def execute(%__MODULE__{} = tk, name, args, ctx \\ nil) do
    case get(tk, name) do
      nil -> %ToolResult{output: "Unknown tool: #{name}", is_error: true}
      %Tool{execute: f} -> f.(args || %{}, ctx || %Context{})
    end
  end

  @doc "Markdown skill catalog for the system prompt."
  @spec skills_prompt(t()) :: String.t()
  def skills_prompt(%__MODULE__{prompt: prompt}), do: prompt

  @doc "Each MCP server's connection status."
  @spec mcp_status(t()) :: %{String.t() => String.t()}
  def mcp_status(%__MODULE__{mcp: nil}), do: %{}
  def mcp_status(%__MODULE__{mcp: mcp}), do: mcp.status

  @doc "OpenAI tool schema for all tools."
  @spec to_openai(t()) :: [map()]
  def to_openai(%__MODULE__{tools: tools}), do: Adapters.to_openai(tools)

  @doc "Anthropic tool schema for all tools."
  @spec to_anthropic(t()) :: [map()]
  def to_anthropic(%__MODULE__{tools: tools}), do: Adapters.to_anthropic(tools)

  @doc "Gemini tool schema for all tools."
  @spec to_gemini(t()) :: [map()]
  def to_gemini(%__MODULE__{tools: tools}), do: Adapters.to_gemini(tools)

  @doc """
  Serve this toolkit as an agent over HTTP (SPEC §7B/§7C). When the A2A
  profile is present (`opts[:a2a]`, or a top-level `a2a` config block the
  toolkit was built from) it mounts the Agent Card + JSON-RPC endpoint
  fulfilled by `opts[:client]`; a message's A2A `contextId` keys the
  conversation via `Client.ask`. When the MCP profile is present
  (`opts[:mcp]`, or a top-level `mcpServer` config block) it mounts a
  streamable-HTTP MCP server at `/mcp` exposing the toolkit's unified tools.
  Returns a stoppable `Toolnexus.Serve.Handle`.
  """
  @spec serve(t(), String.t(), keyword()) :: Toolnexus.Serve.Handle.t()
  def serve(%__MODULE__{} = tk, addr, opts \\ []) do
    a2a = opts[:a2a] || tk.a2a_config
    mcp = opts[:mcp] || tk.mcp_server_config
    client = opts[:client]

    Toolnexus.Serve.start(addr,
      a2a: a2a,
      skills: if(tk.skill, do: tk.skill.skills, else: []),
      run_task: fn text, context_id ->
        # contextId keys the conversation via Client.ask, so a peer's turns are
        # remembered through the client's ConversationStore; nil ⇒ stateless run.
        if context_id,
          do: Toolnexus.Client.ask(client, text, tk, id: context_id),
          else: Toolnexus.Client.run(client, text, tk)
      end,
      on_task: opts[:on_task],
      mcp: mcp,
      mcp_tools: if(mcp, do: tk.tools),
      on_call: opts[:on_call]
    )
  end

  @doc "Disconnect every MCP client."
  @spec close(t()) :: :ok
  def close(%__MODULE__{mcp: nil}), do: :ok
  def close(%__MODULE__{mcp: mcp}), do: Mcp.close(mcp)
end
